src/main/java/org/mule/service/http/impl/service/client/HttpClientConnectionManager.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.service.http.impl.service.client;
+
+import org.mule.runtime.api.scheduler.SchedulerConfig;
+import org.mule.runtime.api.scheduler.SchedulerService;
+import org.mule.runtime.http.api.client.ClientNotFoundException;
+import org.mule.runtime.http.api.client.HttpClient;
+import org.mule.runtime.http.api.client.HttpClientConfiguration;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Manages client connections
+ *
+ * @since 1.1.5
+ */
+public class HttpClientConnectionManager implements ContextHttpClientFactory {
+
+  private final ConcurrentMap<ClientIdentifier, HttpClient> clients = new ConcurrentHashMap<>();
+
+  @Override
+  public HttpClient create(HttpClientConfiguration config,
+                           String context,
+                           SchedulerService schedulerService,
+                           SchedulerConfig schedulerConfig) {
+
+    ClientIdentifier identifier = new ClientIdentifier(context, config.getName());
+    if (clients.containsKey(identifier)) {
```

Comment:
Could there be a race condition here and a client be left in limbo? Check servers too.

Reply:
True that. I did copy this from your original code so shame on you!